File: app/src/main/java/com/pr0gramm/app/ui/IdFragmentStatePagerAdapter.java
package com.pr0gramm.app.ui;

import android.annotation.SuppressLint;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.util.LongSparseArray;
import android.support.v4.view.PagerAdapter;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import com.google.common.base.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.LinkedHashMap;

import static com.google.common.base.Predicates.alwaysTrue;
import static com.google.common.collect.Iterators.limit;
import static com.google.common.collect.Iterators.removeIf;

/**
 * This implementation has a {@link #getItemId(int)} to identify items
 * and fragments, even if they change places between adapter updates.
 *
 * @see android.support.v4.app.FragmentStatePagerAdapter
 */
public abstract class IdFragmentStatePagerAdapter extends PagerAdapter {
    private static final Logger logger = LoggerFactory.getLogger("IdFragmentStatePagerAdapter");

    private static final String TAG = "MyFrgmntStatePgrAdapter";
    private static final boolean DEBUG = false;

    private final FragmentManager mFragmentManager;
    private FragmentTransaction mCurTransaction = null;

    private final LinkedHashMap<Long, Fragment.SavedState> mSavedState = new LinkedHashMap<>();
    private final LongSparseArray<Fragment> mFragments = new LongSparseArray<>();
    private Fragment mCurrentPrimaryItem = null;

    public IdFragmentStatePagerAdapter(FragmentManager fm) {
        mFragmentManager = fm;
    }

    /**
     * Return the Fragment associated with a specified position.
     */
    public abstract Fragment getItem(int position);

    @Override
    public void startUpdate(ViewGroup container) {
        if (container.getId() == View.NO_ID) {
            throw new IllegalStateException("ViewPager with adapter " + this
                    + " requires a view id");
        }
    }

    /**
     * Gets the fragment for the given position, if one already exists.
     *
     * @param position The position of the fragment to get
     */
    public Optional<Fragment> getFragment(int position) {
        long itemId = getItemId(position);
        return Optional.fromNullable(mFragments.get(itemId));
    }

    @SuppressLint("CommitTransaction")
    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        long id = getItemId(position);

        // If we already have this item instantiated, there is nothing
        // to do.  This can happen when we are restoring the entire pager
        // from its saved state, where the fragment manager has already
        // taken care of restoring the fragments we previously had instantiated.
        Fragment f = mFragments.get(id);
        if (f != null) {
            return f;
        }

        if (mCurTransaction == null) {
            mCurTransaction = mFragmentManager.beginTransaction();
        }

        Fragment fragment = getItem(position);
        if (DEBUG) Log.v(TAG, "Adding item #" + position + ": f=" + fragment);
        Fragment.SavedState fss = mSavedState.get(id);
        if (fss != null) {
            fragment.setInitialSavedState(fss);
        }

        fragment.setMenuVisibility(false);
        fragment.setUserVisibleHint(false);
        mFragments.put(id, fragment);
        mCurTransaction.add(container.getId(), fragment);

        return fragment;
    }

    protected abstract long getItemId(int position);

    @SuppressLint("CommitTransaction")
    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        Fragment fragment = (Fragment) object;
        long id = getItemId(position);

        if (mCurTransaction == null) {
            mCurTransaction = mFragmentManager.beginTransaction();
        }
        if (DEBUG) Log.v(TAG, "Removing item #" + position + ": f=" + object
                + " v=" + ((Fragment) object).getView());

        mSavedState.remove(id);

        try {
            mSavedState.put(id, mFragmentManager.saveFragmentInstanceState(fragment));
        } catch (IllegalStateException ignored) {
            // looks like this sometimes happen during save if the frgment is not in the
            // fragment manager. We will ignore it.
        }

        // remove the oldest items
        if (mSavedState.size() > 10) {
            removeIf(limit(mSavedState.values().iterator(), mSavedState.size() - 10), alwaysTrue());
        }

        mFragments.remove(id);

        mCurTransaction.remove(fragment);
    }

    @Override
    public void setPrimaryItem(ViewGroup container, int position, Object object) {
        Fragment fragment = (Fragment) object;
        if (fragment != mCurrentPrimaryItem) {
            if (mCurrentPrimaryItem != null) {
                mCurrentPrimaryItem.setMenuVisibility(false);
                mCurrentPrimaryItem.setUserVisibleHint(false);
            }
            if (fragment != null) {
                try {
                    fragment.setMenuVisibility(true);
                    fragment.setUserVisibleHint(true);
                } catch (NullPointerException ignored) {
                    // another workaround:
                    // support library might access a fragment manager that is null.
                }
            }
            mCurrentPrimaryItem = fragment;
        }
    }

    @Override
    public void finishUpdate(ViewGroup container) {
        try {
            if (mCurTransaction != null) {
                mCurTransaction.commitNowAllowingStateLoss();
                mCurTransaction = null;
            }
        } catch (IllegalStateException ignored) {
            // Sometimes we get a "activity has been destroyed." exception.
        }
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return ((Fragment) object).getView() == view;
    }
//
//    @Override
//    public Parcelable saveState() {
//        Bundle state = null;
//        if (mSavedState.size() > 0) {
//            state = new Bundle();
//
//            long[] ids = new long[mSavedState.size()];
//            Parcelable[] states = new Parcelable[mSavedState.size()];
//
//            ImmutableList<Map.Entry<Long, Fragment.SavedState>> entries = ImmutableList.copyOf(mSavedState.entrySet());
//            for (int idx = 0; idx < entries.size(); idx++) {
//                ids[idx] = entries.get(idx).getKey();
//                states[idx] = entries.get(idx).getValue();
//            }
//            state.putLongArray("ids", ids);
//            state.putParcelableArray("states", states);
//        }
//
//        for (int idx = 0; idx < mFragments.size(); idx++) {
//            Fragment f = mFragments.valueAt(idx);
//            if (f != null) {
//                if (state == null) {
//                    state = new Bundle();
//                }
//                String key = "f" + mFragments.keyAt(idx);
//                mFragmentManager.putFragment(state, key, f);
//            }
//        }
//        return state;
//    }
//
//    @Override
//    public void restoreState(Parcelable state, ClassLoader loader) {
//        if (state != null) {
//            Bundle bundle = (Bundle) state;
//            bundle.setClassLoader(loader);
//
//            long[] ids = bundle.getLongArray("ids");
//            Parcelable[] fss = bundle.getParcelableArray("states");
//
//            mSavedState.clear();
//            mFragments.clear();
//            if (fss != null) {
//                for (int i = 0; i < fss.length; i++) {
//                    mSavedState.put(ids[i], (Fragment.SavedState) fss[i]);
//                }
//            }
//
//            Iterable<String> keys = bundle.keySet();
//            for (String key : keys) {
//                if (key.startsWith("f")) {
//                    long id = Long.parseLong(key.substring(1));
//                    Fragment f = mFragmentManager.getFragment(bundle, key);
//                    if (f != null) {
//                        f.setMenuVisibility(false);
//                        mFragments.put(id, f);
//                    } else {
//                        logger.warn("Bad fragment at key " + key);
//                    }
//                }
//            }
//        }
//    }
}

File: app/src/main/java/com/pr0gramm/app/ui/IdFragmentStatePagerAdapter.kt
package com.pr0gramm.app.ui

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Parcelable
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.viewpager.widget.PagerAdapter
import com.pr0gramm.app.util.Logger
import com.pr0gramm.app.util.LongSparseArray
import java.util.*

/**
 * This implementation has a [.getItemId] to identify items
 * and fragments, even if they change places between adapter updates.

 * @see androidx.fragment.app.FragmentStatePagerAdapter
 */
abstract class IdFragmentStatePagerAdapter(private val mFragmentManager: FragmentManager) : PagerAdapter() {
    private val logger = Logger("IdFragmentStatePagerAdapter")

    // we only cache the most recent few saved states.
    private val mSavedState = object : LinkedHashMap<Long, Fragment.SavedState>() {
        override fun removeEldestEntry(eldest: Map.Entry<Long, Fragment.SavedState>): Boolean {
            return size > 5
        }
    }

    private val mFragments = LongSparseArray<Fragment>()
    private var mCurTransaction: FragmentTransaction? = null
    private var mCurrentPrimaryItem: Fragment? = null

    /**
     * Return the Fragment associated with a specified position.
     */
    abstract fun getItem(position: Int): Fragment

    override fun startUpdate(container: ViewGroup) {
        if (container.id == View.NO_ID) {
            throw IllegalStateException("ViewPager with adapter ${this} requires a view id")
        }
    }

    /**
     * Gets the fragment for the given position, if one already exists.

     * @param position The position of the fragment to get
     */
    fun getFragment(position: Int): Fragment? {
        val itemId = getItemId(position)
        return mFragments[itemId]
    }

    @SuppressLint("CommitTransaction")
    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val id = getItemId(position)

        // If we already have this item instantiated, there is nothing
        // to do.  This can happen when we are restoring the entire pager
        // from its saved state, where the fragment manager has already
        // taken care of restoring the fragments we previously had instantiated.
        val f = mFragments[id]
        if (f != null) {
            return f
        }

        if (mCurTransaction == null) {
            mCurTransaction = mFragmentManager.beginTransaction()
        }

        val fragment = getItem(position)
        val fss = mSavedState[id]
        if (fss != null) {
            fragment.setInitialSavedState(fss)
        }

        fragment.setMenuVisibility(false)
        fragment.userVisibleHint = false
        mFragments.put(id, fragment)
        mCurTransaction?.add(container.id, fragment)

        return fragment
    }

    protected abstract fun getItemId(position: Int): Long

    @SuppressLint("CommitTransaction")
    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        val fragment = `object` as Fragment

        if (mCurTransaction == null) {
            mCurTransaction = mFragmentManager.beginTransaction()
        }

        try {
            val id = getItemId(position)
            mFragments.remove(id)

            mSavedState.remove(id)
            mFragmentManager.saveFragmentInstanceState(fragment)?.let { state ->
                mSavedState.put(id, state)
            }
        } catch (ignored: Exception) {
            // looks like this sometimes happen during save if the fragment is not in the
            // fragment manager. We will ignore it.
        }

        mCurTransaction?.remove(fragment)
    }

    override fun setPrimaryItem(container: ViewGroup, position: Int, `object`: Any) {
        val fragment = `object` as Fragment
        if (fragment !== mCurrentPrimaryItem) {
            mCurrentPrimaryItem?.let {
                it.setMenuVisibility(false)
                it.userVisibleHint = false
            }

            try {
                fragment.setMenuVisibility(true)
                fragment.userVisibleHint = true
            } catch (ignored: NullPointerException) {
                // another workaround:
                // support library might access a fragment manager that is null.
            }

            mCurrentPrimaryItem = fragment
        }
    }

    override fun finishUpdate(container: ViewGroup) {
        try {
            mCurTransaction?.commitNowAllowingStateLoss()
            mCurTransaction = null
        } catch (ignored: RuntimeException) {
            // Sometimes we get a "activity has been destroyed." exception.
        }
    }

    override fun isViewFromObject(view: View, obj: Any): Boolean {
        return (obj as Fragment).view === view
    }

    override fun saveState(): Parcelable? {
        var state: Bundle? = null

        if (mSavedState.size > 0) {
            val ids = LongArray(mSavedState.size)
            val states = arrayOfNulls<Parcelable>(mSavedState.size)

            mSavedState.entries.forEachIndexed { idx, entry ->
                ids[idx] = entry.key
                states[idx] = entry.value
            }

            state = Bundle()
            state.putLongArray("ids", ids)
            state.putParcelableArray("states", states)
        }

        for (idx in 0 until mFragments.size) {
            state = state ?: Bundle()
            val key = "f" + mFragments.keyAt(idx)

            try {
                mFragmentManager.putFragment(state, key, mFragments.valueAt(idx))
            } catch (err: IllegalStateException) {
                logger.warn("Could not put fragment into the bundle. Skipping.", err)
            }
        }

        return state
    }

    override fun restoreState(state: Parcelable?, loader: ClassLoader?) {
        if (state is Bundle) {
            state.classLoader = loader

            val ids = state.getLongArray("ids")
            val fss = state.getParcelableArray("states")

            mFragments.clear()
            mSavedState.clear()

            if (ids != null && fss != null) {
                for (i in fss.indices) {
                    mSavedState.put(ids[i], fss[i] as Fragment.SavedState)
                }
            }

            val keys = state.keySet()
            for (key in keys) {
                if (key.startsWith("f")) {
                    val id = java.lang.Long.parseLong(key.substring(1))
                    val f = mFragmentManager.getFragment(state, key)
                    if (f != null) {
                        f.setMenuVisibility(false)
                        mFragments.put(id, f)
                    } else {
                        logger.warn { "Bad fragment at key $key" }
                    }
                }
            }
        }
    }
}
